### DISCUSSION TITLE
Possible to expose allowed send parameters for a state?

### DISCUSSION THREAD

**Discussion by pke** [statelyai/xstate#2652] [Q&A]
Lets say I have a state that collects all letters of the alphabet but allows as input only letters that it has not already collected.

The UI using the statemachine should not know anything about the alphabet. Instead it should be able to ask the statemachine:
1. Which actions are allowed now (xstate supports that already)
2. Which aguments each action allows (this seems not supported)

And with that the statemachine could instruct the UI to allow the "COLLECT" action with one of the not yet sent letters of the alphabet:

Sending this action would remove the letter `a` from the potential list of allowed `letter` values for the `COLLECT` action:

`{ type: "COLLECT", letter: "a" }`

My poor mans approach was to put this info into the context for now: 

```
{
  actions: [
  {
     type: "COLLECT",
     fields: [
        { name:  "letter", value: "b" }
     ]
  },
  {
     type: "COLLECT",
     fields: [
        { name:  "letter", value: "c" }
     ]
  }
  // ...
  ]
}
```

Maybe the Model API could help somehow in that regard? TypeScript is of no help, because it does not exist during runtime. But for that feature we need runtime readable "type" informations.

**Comment by Andarist** [comment]
> And with that the statemachine could instruct the UI to allow the "COLLECT" action with one of the not yet sent letters of the alphabet

Doesn't this mean that your requirements are dynamic and cant be easily encoded in the config object?

**Reply by pke** [reply]
Yes. That's why this can't be done in meta data.
